File: Longest Substring Without Repeating Characters - Leetcode 3/Longest Substring Without Repeating Characters - Leetcode 3.cpp
#include <unordered_set>
#include <algorithm>
using namespace std;



class Solution {
public:
    int lengthOfLongestSubstring(const string s) {

        int longest = 0;

        for (auto i = 0; i < s.size(); ++i) {
            for (auto substr_len = 1; i + substr_len <= s.size(); ++substr_len) {
                unordered_set<char> seen{s.cbegin() + i, s.cbegin() + i + substr_len};
                if (seen.size() == substr_len) 
                    longest = max(longest, substr_len);
            }
        }

        return longest;
    }
};
// Time Complexity: O(n^3)
// Space Complexity: O(n)
    


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;
        int l = 0, longest = 0;

        for (int r = 0; r < s.size(); r++) {
            while (set.find(s[r]) != set.end()) {
                set.erase(s[l]);
                l++;
            }

            longest = max(longest, r - l + 1);
            set.insert(s[r]);
        }

        return longest;
    }
};

File: Longest Substring Without Repeating Characters - Leetcode 3/Longest Substring Without Repeating Characters - Leetcode 3.py
# Brute Force
def length_of_longest_substring(s: str) -> int:
    max_length = 0

    for i in range(len(s)):
        for j in range(i, len(s)):
            if s[j] in s[i:j]:  # Check if the character already appeared in the substring
                break
            max_length = max(max_length, j - i + 1)

    return max_length
    # Time: O(n^3)
    # Space: O(n)

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        longest = 0
        sett = set()
        n = len(s)

        for r in range(n):
            while s[r] in sett:
                sett.remove(s[l])
                l += 1

            w = (r - l) + 1
            longest = max(longest, w)
            sett.add(s[r])

        return longest

# Time Complexity: O(n)
# Space Complexity: O(n) (but because we're limited to common characters only, it's actually O(1))
